Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs

```diff
@@ -147,37 +161,468 @@ bool loadCanceled
-      private TextDocumentItem document { get; init; }
+      private ICompilationStatusNotificationPublisher publisher { get; }
+      private DafnyDocument document { get; }
 
-      public VerificationProgressReporter(TextDocumentItem document,
-                                          ICompilationStatusNotificationPublisher publisher) {
+      private ILogger<VerificationProgressReporter> logger { get; }
+      private IDiagnosticPublisher diagnosticPublisher { get; }
+
+      public VerificationProgressReporter(ILogger<VerificationProgressReporter> logger,
+                                          DafnyDocument document,
+                                          ICompilationStatusNotificationPublisher publisher,
+                                          IDiagnosticPublisher diagnosticPublisher
+      ) {
         this.document = document;
         this.publisher = publisher;
+        this.logger = logger;
+        this.diagnosticPublisher = diagnosticPublisher;
       }
 
+      /// <summary>
+      /// Sends a more precise verification status message to the client's status bar
+      /// </summary>
       public void ReportProgress(string message) {
-        publisher.SendStatusNotification(document, CompilationStatus.VerificationStarted, message);
+        publisher.SendStatusNotification(document.Text, CompilationStatus.VerificationStarted, message);
+      }
+
+      /// <summary>
+      /// Fills up the document with empty verification diagnostics, one for each top-level declarations
+      /// Possibly migrates previous diagnostics
+      /// </summary>
+      public void RecomputeVerificationTree() {
+        var previousTrees = document.VerificationTree.Children;
+
+        List<VerificationTree> result = new List<VerificationTree>();
+
+        void AddAndPossiblyMigrateVerificationTree(VerificationTree verificationTree) {
+          var position = verificationTree.Position;
+          var previousTree = previousTrees.FirstOrDefault(
+            oldNode => oldNode != null && oldNode.Position == position,
+            null);
+          if (previousTree != null) {
+            previousTree.SetObsolete();
+            verificationTree.StatusVerification = previousTree.StatusVerification;
+            verificationTree.StatusCurrent = CurrentStatus.Obsolete;
+            verificationTree.Children = previousTree.Children;
+          }
+          result.Add(verificationTree);
+        }
+
+        var documentFilePath = document.GetFilePath();
+        foreach (var module in document.Program.Modules()) {
+          foreach (var topLevelDecl in module.TopLevelDecls) {
+            if (topLevelDecl is TopLevelDeclWithMembers topLevelDeclWithMembers) {
+              foreach (var member in topLevelDeclWithMembers.Members) {
+                if (member.tok.filename != documentFilePath) {
+                  continue;
+                }
+                if (member is Field) {
+                  if (member.BodyEndTok.line == 0) {
```

Comment:
Looking deeper, it currently means that there is no body, so the constant does not have a value. See here:
https://github.com/dafny-lang/dafny/blob/verification-diagnostics-core/Source/Dafny/Dafny.atg#L1374
However, the goal of this if was to prevent the `GetLspRange` next line to raise an exception

Reply:
Would it be possible to replace `if (member is Field) { if (member.BodyEndTok.line == 0).. ` with `if (member is ConstantField)` ?

Alternatively, you could extract this into a variable to explain the condition, like `var constantHasNoBody = member.BodyEndTok.line == 0`.

Reply:
Since that ConstantFields can have bodies, I'll go the second route.

Reply:
> Since that ConstantFields can have bodies, I'll go the second route.

My impression was that only ConstantFields have bodies and they always have them, so if you check for `ConstantField1` you no longer need another check. Did I misread?

Reply:
You can declare Constant fields without a body. it's just that the compiler won't accept it, but it works for verification purposes, I think.